be2-scala/src/main/scala/ch/epfl/pop/pubsub/PubSubMediator.scala

```diff
@@ -3,62 +3,125 @@ package ch.epfl.pop.pubsub
-              log.info(reason)
-              senderRef ! SubscribeToNAck(channel, reason)
-            } else {
-              log.info(s"Subscribing $senderRef to channel '$channel'")
-              set += senderRef
-              senderRef ! SubscribeToAck(channel)
-            }
-          case _ =>
-            log.info(s"Subscribing $senderRef to channel '$channel'")
-            channelMap = channelMap + (channel -> mutable.Set(senderRef))
-            senderRef ! SubscribeToAck(channel)
-        }
+  private def subscribeTo(channel: Channel, clientActorRef: ActorRef): Future[PubSubMediatorMessage] = {
+    val ask: Future[PubSubMediatorMessage] = (dbActor ? DbActor.ChannelExists(channel)).map {
+      // if the channel exists in db, we can start thinking about subscribing a client to it
+      case DbActorAck() => channelMap.get(channel) match {
+        // if we have people already subscribed to said channel
+        case Some(set) =>
+          if (set.contains(clientActorRef)) {
+            val reason: String = s"Actor $clientActorRef was already subscribed to channel '$channel'"
+            log.info(reason)
+            SubscribeToNAck(channel, reason)
+          } else {
+            log.info(s"Subscribing $clientActorRef to channel '$channel'")
+            set += clientActorRef
+            SubscribeToAck(channel)
+          }
+
+        // if we have no one subscribed to said channel
         case _ =>
-          val reason: String = s"Channel '$channel' doesn't exist in db"
-          log.info(reason)
-          senderRef ! SubscribeToNAck(channel, reason)
+          log.info(s"Subscribing $clientActorRef to channel '$channel'")
+          channelMap = channelMap + (channel -> mutable.Set(clientActorRef))
+          SubscribeToAck(channel)
       }
 
-    case UnsubscribeFrom(channel) => channelMap.get(channel) match {
-      case Some(set) if set.contains(sender) =>
-        log.info(s"Unsubscribing $sender from channel '$channel'")
-        set -= sender
-        sender ! UnsubscribeFromAck(channel)
-      case Some(_) =>
-        sender ! UnsubscribeFromNAck(channel, s"Actor $sender is not subscribed to channel '$channel'")
+      // db doesn't recognize the channel, thus mediator cannot subscribe anyone to a non existing channel
+      case _ =>
+        val reason: String = s"Channel '$channel' doesn't exist in db"
+        log.info(reason)
+        SubscribeToNAck(channel, reason)
+    }
+
+    ask
+  }
+
+  private def unsubscribeFrom(channel: Channel, clientActorRef: ActorRef): PubSubMediatorMessage = channelMap.get(channel) match {
+    case Some(set) if set.contains(clientActorRef) =>
+      log.info(s"Unsubscribing $clientActorRef from channel '$channel'")
+      set -= clientActorRef
+      UnsubscribeFromAck(channel)
+    case Some(_) =>
+      val reason: String = s"Actor $clientActorRef is not subscribed to channel '$channel'"
+      log.info(reason)
+      UnsubscribeFromNAck(channel, reason)
```

Comment:
I'm wondering, should this be a NAck ?
In practice the system could say "sure, you're unsubscribed" at this point.